File: test/medium/remove_duplicate_letters_test.dart
import 'package:test/test.dart';
import 'package:dark_coding_challenges/medium/remove_duplicate_letters.dart';

class RemoveDuplicateLettersTestCases {
  String s;
  String expected;
  RemoveDuplicateLettersTestCases(this.s, this.expected);
}

void main() {
  group("Feature:", () {
    final testCases = [
      RemoveDuplicateLettersTestCases("bcabc", "abc"),
      RemoveDuplicateLettersTestCases("cbacdcbc", "acdb"),
      RemoveDuplicateLettersTestCases("a", "a"),
      RemoveDuplicateLettersTestCases("aa", "a"),
      RemoveDuplicateLettersTestCases("aaa", "a"),
      RemoveDuplicateLettersTestCases("ab", "ab"),
      RemoveDuplicateLettersTestCases("ba", "ba"),
      RemoveDuplicateLettersTestCases("abc", "abc"),
      RemoveDuplicateLettersTestCases("cba", "cba"),
      RemoveDuplicateLettersTestCases("abab", "ab"),
      RemoveDuplicateLettersTestCases("baba", "ab"),
      RemoveDuplicateLettersTestCases("bbcaac", "bac"),
      RemoveDuplicateLettersTestCases("cdadabcc", "adbc"),
      RemoveDuplicateLettersTestCases("ecbacba", "eacb"),
      RemoveDuplicateLettersTestCases("leetcode", "letcod"),
      RemoveDuplicateLettersTestCases("bcab", "abc"),
      RemoveDuplicateLettersTestCases("zxyxz", "xyz"),
      RemoveDuplicateLettersTestCases("abcdefghijklmnopqrstuvwxyz", "abcdefghijklmnopqrstuvwxyz"),
      RemoveDuplicateLettersTestCases("zyxwvutsrqponmlkjihgfedcba", "zyxwvutsrqponmlkjihgfedcba"),
      RemoveDuplicateLettersTestCases("aabbccddeeff", "abcdef"),
      RemoveDuplicateLettersTestCases("abababab", "ab"),
      RemoveDuplicateLettersTestCases("bcbcbc", "bc"),
      RemoveDuplicateLettersTestCases("cccbbaaa", "cba"),
      RemoveDuplicateLettersTestCases("dacbacbd", "acbd"),
      RemoveDuplicateLettersTestCases("aaaaabbbbbccccc", "abc"),
    ];

    for (final step in testCases) {
      final s = step.s;
      final expected = step.expected;
      test("\n\tGiven: the string $s.\n\tThen: the lexicographical substring is $expected.", () {
        String actual = removeDuplicateLetters(s);
        expect(actual, equals(expected));
      });
    }
  });
}

File: medium/remove_duplicate_letters_test.go
package medium

import (
	"testing"
)

func TestRemoveDuplicateLetters(t *testing.T) {
	cases := []struct {
		s        string
		expected string
	}{

		{"bcabc", "abc"},
		{"cbacdcbc", "acdb"},
		{"a", "a"},
		{"aa", "a"},
		{"aaa", "a"},
		{"ab", "ab"},
		{"ba", "ba"},
		{"abc", "abc"},
		{"cba", "cba"},
		{"abab", "ab"},
		{"bbcaac", "bac"},
		{"ecbacba", "eacb"},
		{"leetcode", "letcod"},
		{"cdadabcc", "adbc"},
		{"bcab", "bca"},
		{"cbac", "bac"},
		{"zxyxz", "xyz"},
		{"aaaaabbbbbccccc", "abc"},
		{"abcdefghijklmnopqrstuvwxyz", "abcdefghijklmnopqrstuvwxyz"},
		{"zyxwvutsrqponmlkjihgfedcba", "zyxwvutsrqponmlkjihgfedcba"},
		{"bcbcbcbc", "bc"},
		{"abacb", "abc"},
		{"aabbcc", "abc"},
		{"ccbbaa", "cba"},
		{"abababab", "ab"},
		{"abcabcabc", "abc"},
		{"edebbed", "bed"},
		{"thesqtitxyetpxloeevdeqifkz", "hesitxyplovdqfkz"},
	}
	for _, c := range cases {
		t.Run("", func(t *testing.T) {
			result := RemoveDuplicateLetters(c.s)
			if result != c.expected {
				t.Errorf("RemoveDuplicateLetters(%q) == %q, expected %q", c.s, result, c.expected)
			}

		})

	}
}
